src/dmv.py

```diff
@@ -34,66 +62,192 @@ class CA_DMV():
+    ]]
+    MUST_FAIL_PATTERNS = [(re.compile(pattern), reason) for pattern, reason in [
+        ("//", "Two half-spaces cannot be in a row."),
+    ]]
+    NORMALIZE_PATTERNS = [(re.compile(pattern), repl) for pattern, repl in [
+        ("[ /]", ""), # Spaces aren't considered for whether or not something is taken.
+    ]]
+
     SUCCESS_PHRASE = "id=\"PersonalizedFormBean\""
     FAILURE_PHRASE = "id=\"plate-configurator\""
 
-    def __init__(self):
+    ERROR_TAKEN = "The license plate number you have selected is no longer available."
+    ERROR_INVALID = "Your license plate request contains invalid characters"
+    ERROR_SHORT = "Your license plate must have at least 2 characters."
+    ERROR_HALFSPACES = "Your license plate number cannot contain 2 half spaces (/) in a row."
+
+    APPNAME = "vanity"
+    CACHE_FILE = "ca_dmv.json"
+
+    def __init__(self, cache_option = CacheOption.DEFAULT):
         self.session = None
+        self.cache = None
+        self.cache_option = cache_option
+        if self.cache_option == CacheOption.SKIP_MEM_CACHE:
+            logger.debug("Not initializing cache due to option %s", self.cache_option)
+        if self.cache_option == CacheOption.CLEAR_DISK_CACHE:
+            self.clear_cache()
 
     def initialize(self):
-        self.check_initialized(force = True)
+        self.check_cache_initialized(force = True)
+        self.check_session_initialized(force = True)
 
-    def checkPlate(self, word):
-        self.check_initialized(force = False)
-        data = CA_DMV.buildPlateRequestData(word)
-        if not data:
-            print("Invalid input, can't check the DMV.")
+    def check_plate(self, word):
+        logger.debug("Checking word: '%s'", word)
+        normalized = CA_DMV.normalize_word(word)
+        if normalized is None:
+            logger.error("Invalid input, can't check the DMV.")
             return False
+        if normalized != word:
+            logger.debug("Normalized word to: '%s'", normalized)
+        cached = self.cache_read(normalized)
+        if cached is not None:
+            logger.debug("Found result in cache: ('%s': %s)", normalized, cached)
+            return cached
+        self.check_session_initialized()
+        data = CA_DMV.build_plate_request_data(normalized)
         result = self.session.post(
-            url = CA_DMV.dmvUrl(CA_DMV.PATH_PLATE), 
+            url = CA_DMV.dmv_url(CA_DMV.PATH_PLATE),
             data = data,
         )
-        available = CA_DMV.checkPlateResult(result)
+        available = CA_DMV.check_plate_response(result)
         if available is None:
-            print("Got an unknown result for a plate, neither success nor failure!")
+            logger.error("Got an unknown result for a plate, neither success nor failure!")
             return False
+        logger.debug("Got result from API: ('%s': %s)", normalized, available)
+        self.cache_write(normalized, available)
+        self.commit_cache()
```

Comment:
Seems expensive to do for every plate

Reply:
Yup, I have that as future improvement:
> The in memory cache currently commits to disk on every change, this can be optimized.

Writing to disk is less expensive than the network call though (hopefully) so it's not as bad as it looks - but yeah, ideally these writes would be batched.